actors/eam/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+        subaddress: new_addr.0.to_vec().into(),
+    };
+
+    let ret: ext::init::Exec4Return = rt
+        .send(
+            &INIT_ACTOR_ADDR,
+            ext::init::EXEC4_METHOD,
+            RawBytes::serialize(&init_params)?,
+            rt.message().value_received(),
+        )?
+        .deserialize()?;
+
+    Ok(RawBytes::serialize(EamReturn::from_exec4(ret, new_addr))?)
+}
+
+/// lookup caller's raw ETH address
+fn get_caller_address<BS, RT>(rt: &RT) -> Result<EthAddress, ActorError>
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    let caller_id = rt.message().caller().id().unwrap();
+
+    let addr = rt.lookup_address(caller_id);
+
+    match addr.map(|a| *a.payload()) {
+        Some(Payload::Delegated(eth)) => {
+            if eth.namespace() == EAM_ACTOR_ID {
+                Ok(EthAddress(eth.subaddress().try_into().unwrap()))
+            } else {
+                Err(ActorError::assertion_failed("Caller is not in the EVM namespace.".into()))
+            }
+        }
+        _ => Err(ActorError::assertion_failed(
+            "All FEVM actors should have a delegated address.".to_string(),
+        )),
+    }
+}
+
+pub struct EamActor;
+impl EamActor {
+    pub fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        let actor_id = rt.resolve_address(&rt.message().receiver()).unwrap();
+        if actor_id != EAM_ACTOR_ID {
+            return Err(ActorError::forbidden(format!(
+                "The Ethereum Address Manager must be deployed at {EAM_ACTOR_ID}, was deployed at {actor_id}"
+            )));
+        }
+        rt.validate_immediate_caller_is(std::iter::once(&INIT_ACTOR_ADDR))
+    }
+
+    pub fn create<BS, RT>(rt: &mut RT, params: CreateParams) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_type(iter::once(&Type::EVM))?;
+        assert_code_size(&params.initcode)?;
+
+        let caller_addr = get_caller_address(rt)?;
+        // CREATE logic
+        let rlp = RlpCreateAddress { address: caller_addr, nonce: params.nonce };
+        let eth_addr = EthAddress(hash_20(rt, &rlp.rlp_bytes()));
+
+        // send to init actor
+        create_actor(rt, caller_addr, eth_addr, params.initcode)
+    }
+
+    pub fn create2<BS, RT>(rt: &mut RT, params: Create2Params) -> Result<RawBytes, ActorError>
+    where
+        BS: Blockstore + Clone,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_type(iter::once(&Type::EVM))?;
+        assert_code_size(&params.initcode)?;
+
+        // CREATE2 logic
+        let inithash = rt.hash(SupportedHashes::Keccak256, &params.initcode);
+
+        let caller_addr = get_caller_address(rt)?;
+
+        let eth_addr = EthAddress(hash_20(
+            rt,
+            &[&[0xff], caller_addr.0.as_slice(), &params.salt, &inithash].concat(),
```

Comment:
```suggestion
            &[&[0xff], &caller_addr.0[..], &params.salt, &inithash].concat(),
```

(slightly more idiomatic... but doesn't really matter. feel free to ignore)